node/router/src/heartbeat.rs

```diff
@@ -148,18 +158,33 @@ pub trait Heartbeat<N: Network>: Outbound<N> {
             // Initialize an RNG.
             let rng = &mut OsRng;
 
-            // TODO (howardwu): As a validator, prioritize disconnecting from clients and provers.
+            // Determine the provers to disconnect from.
+            let prover_ips_to_disconnect = self
+                .router()
+                .connected_provers()
+                .into_iter()
+                .filter(|peer_ip| !trusted.contains(peer_ip) && !bootstrap.contains(peer_ip))
+                .choose_multiple(rng, num_surplus_provers);
+
+            // TODO (howardwu): As a validator, prioritize disconnecting from clients.
             //  Remove RNG, pick the `n` oldest nodes.
-            // Determine the peers to disconnect from.
+            // Determine the clients and validators to disconnect from.
             let peer_ips_to_disconnect = self
                 .router()
-                .connected_peers()
+                .get_connected_peers()
                 .into_iter()
-                .filter(|peer_ip| !trusted.contains(peer_ip) && !bootstrap.contains(peer_ip))
-                .choose_multiple(rng, num_surplus);
+                .filter_map(|peer| {
+                    let peer_ip = peer.ip();
+                    if !trusted.contains(&peer_ip) && !bootstrap.contains(&peer_ip) && !peer.is_prover() {
```

Comment:
```suggestion
                    if !peer.is_prover() && !trusted.contains(&peer_ip) && !bootstrap.contains(&peer_ip) {
```
for faster short-circuiting

Reply:
02a08ec20